text/0000-capture-disjoint-fields.md

```diff
@@ -0,0 +1,196 @@
+    self.list.retain(|i| self.filter.allowed(i));
+}
+```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The borrow checker understands structs sufficiently to know that it's possible
+to borrow disjoint fields of a struct simultaneously. Structs can also be
+destructed and moved piece-by-piece. This functionality should be available
+anywhere, including from within closures:
+
+```rust
+struct OneOf {
+    text: String,
+    of: Vec<String>,
+}
+
+impl OneOf {
+    pub fn matches(self) -> bool {
+        // Ok! destructure self
+        self.of.into_iter().any(|s| s == self.text)
+    }
+
+    pub fn filter(&mut self) {
+        // Ok! mutate and inspect self
+        self.of.retain(|s| s != &self.text)
+    }
+}
+```
+
+Rust will prevent dangerous double usage:
+
+```rust
+struct FirstDuplicated(Vec<String>)
+
+impl FirstDuplicated {
+    pub fn first_count(self) -> usize {
+        // Error! can't destructure and mutate same data
+        self.0.into_iter()
+            .filter(|s| &s == &self.0[0])
+            .count()
+    }
+
+    pub fn remove_first(&mut self) {
+        // Error! can't mutate and inspect same data
+        self.0.retain(|s| s != &self.0[0])
+    }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+There exists an nonoptimal but trivial desugar/workaround that covers all cases
+via the following expansion:
+```
+capture := [|'&'|'&mut '] ident ['.' ident]*
+
+'|' args '|' [$e($c:capture):expression]* =>
+'{'
+    ['let' $name:ident '=' $c ';']*
+    '|' args '|' [$e($name)]*
+'}'
+```
+
+Applied to the first two examples:
+```rust
+let _a = &mut foo.a;
+let b = &mut foo.b;
+|| b;
+```
+```rust
+let _a = &mut foo.a;
+let b = foo.b;
+move || b;
+```
+
+This proves that the RFC can be safely implemented without violating any
+existing assumptions. Also, because the compiler would become strictly more
+lenient, it is nonbreaking.
+
+This RFC should not be implemented as such a desugar. Rather, the two following
+changes might be made:
+- Borrowck rules are altered so that capture-by-reference allows simultaneous
+  borrowing of disjoint fields.
+- Codegen and borrowck are altered so that move closures capture only used
+  fields.
+
+## Examples of an ideal implementation
+
+```rust
+|| &mut foo.a;
+```
+- Borrowck passes because `foo` is not borrowed elsewhere.
+- The closure captures a pointer to `foo.a`.
+
+```rust
+let _a = &mut foo.a;
+|| &mut foo.b;
+```
+- Borrowck passes because `foo.a` and `foo.b` are disjoint.
+- The closure captures a pointer to `foo.b`.
+
+```rust
+let _a = &mut foo.a;
+|| (&mut foo.b, &mut foo.c);
+```
+- Borrowck passes because `foo.a`, `foo.b`, and `foo.c` are disjoint.
+- The closure captures a pointer to `foo`.
+
+```rust
+move || foo.a;
+```
+- Borrowck passes because `foo` is not borrowed elsewhere.
+- The closure moves and captures `foo.a` but not `foo.b`.
```

Comment:
This won't work if `foo`'s type implements `Drop`

Reply:
@oli-obk Thanks for catching that! For those examples I am assuming only cases where the desugar will compile, since it should be equivalent to the final implementation. If `foo` implements `Drop` then it can not be destructed anyway, capturing or not.